nanobenchmarks/main.cpp

```diff
@@ -0,0 +1,268 @@
+ABSL_FLAG(double, input, 5, "Input for the benchmarked functions");
+
+// Adding support for flag types only works using ADL (or by being in
+// marshalling.h), so we do this, which is UB.
+namespace std {
+
+bool AbslParseFlag(absl::string_view const text,
+                   std::vector<double>* const flag,
+                   std::string* const error) {
+  flag->clear();
+  for (absl::string_view const element : absl::StrSplit(text, ',')) {
+    if (!absl::ParseFlag(element, &flag->emplace_back(), error)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::string AbslUnparseFlag(std::vector<double> const& flag) {
+  return absl::StrJoin(flag, ",");
+}
+
+}  // namespace std
+
+ABSL_FLAG(std::vector<double>,
+          quantiles,
+          (std::vector<double>{0.001, 0.01, 0.05, 0.1, 0.25, 0.5}),
+          "Quantiles to report");
+
+namespace principia {
+namespace nanobenchmarks {
+namespace _main {
+namespace {
+
+using namespace principia::mathematica::_logger;
+using namespace principia::mathematica::_mathematica;
+using namespace principia::nanobenchmarks::_function_registry;
+using namespace principia::nanobenchmarks::_microarchitectures;
+using namespace principia::nanobenchmarks::_performance_settings_controller;
+using namespace principia::testing_utilities::_statistics;
+
+struct LatencyDistributionTable {
+  double min;
+  std::vector<double> quantiles;
+
+  static std::string const& heading() {
+    static std::string const& result = [] {
+      std::stringstream& out = *new std::stringstream();
+      std::print(out, "{:>8}", "min");
+      for (double const q : absl::GetFlag(FLAGS_quantiles)) {
+        if (q < 1e-3) {
+          std::print(out, "{:>7}‱", 10'000 * q);
+        } else if (q < 1e-2) {
+          std::print(out, "{:>7}‰", 1000 * q);
+        } else {
+          std::print(out, "{:>7}%", 100 * q);
+        }
+      }
+      return out.str();
+    }();
+    return result;
+  }
+
+  std::string Row() const {
+    std::stringstream out;
+    std::print(out, "{:8.2f}", min);
+    for (double const quantile : quantiles) {
+      std::print(out, "{:+8.2f}", quantile - min);
+    }
+    return out.str();
+  }
+};
+
+LatencyDistributionTable operator*(double a, LatencyDistributionTable x) {
+  LatencyDistributionTable result{a * x.min};
+  for (double const quantile : x.quantiles) {
+      result.quantiles.push_back(a * quantile);
+  }
+  return result;
+}
+
+LatencyDistributionTable operator+(LatencyDistributionTable x, double b) {
+  LatencyDistributionTable result{x.min + b};
+  for (double const quantile : x.quantiles) {
+      result.quantiles.push_back(quantile + b);
+  }
+  return result;
+}
+
+__declspec(noinline) LatencyDistributionTable
+    Benchmark(BenchmarkedFunction f, Logger* logger) {
+  std::size_t const sample_count = absl::GetFlag(FLAGS_samples);
+  std::size_t const loop_iterations = absl::GetFlag(FLAGS_loop_iterations);
+  static std::vector<double>& samples = *new std::vector<double>(
```

Comment:
Let's use a pointer, not a reference, to avoid depending on lifetime extension.

Reply:
Lifetime is obvious : *new, we leak.
Let’s not litter the code with (*) below.